File: simulation/swarm/sensors/py/ideal_sensor_test.py
import numpy as np
from absl.testing import absltest
from simulation.swarm.proto.state_pb2 import State

from simulation.swarm.py.model_agent import ModelAgent
from simulation.swarm.sensors.py.ideal_sensor import IdealSensor


class IdealSensorTargetAtBoresightTestCase(absltest.TestCase):

    def setUp(self):
        # Configure the agent.
        agent_state = State()
        self.agent_velocity = np.array([0, 4, 0])
        (
            agent_state.velocity.x,
            agent_state.velocity.y,
            agent_state.velocity.z,
        ) = self.agent_velocity
        self.agent = ModelAgent(agent_state)
        self.sensor = IdealSensor(self.agent)

        # Configure the target.
        target_state = State()
        self.target_position = np.array([0, 4, 0])

        (
            target_state.position.x,
            target_state.position.y,
            target_state.position.z,
        ) = self.target_position
        self.target_velocity = np.array([2, 2, -1])
        (
            target_state.velocity.x,
            target_state.velocity.y,
            target_state.velocity.z,
        ) = self.target_velocity
        self.target = ModelAgent(target_state)

    def test_sense_position_range(self):
        expected_range = np.linalg.norm(self.target_position)
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.range,
            expected_range,
        )

    def test_sense_position_azimuth(self):
        expected_azimuth = np.arctan(self.target_position[0] /
                                     self.target_position[1])
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.azimuth,
            expected_azimuth,
        )

    def test_sense_position_elevation(self):
        expected_elevation = np.arctan(
            self.target_position[2] /
            np.sqrt(self.target_position[0]**2 + self.target_position[1]**2))
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.elevation,
            expected_elevation,
        )

    def test_sense_velocity_range(self):
        expected_range_rate = -2
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.range,
            expected_range_rate,
        )

    def test_sense_velocity_azimuth(self):
        expected_azimuth_velocity = 2 / 4
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.azimuth,
            expected_azimuth_velocity,
        )

    def test_sense_velocity_elevation(self):
        expected_elevation_velocity = -1 / 4
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.elevation,
            expected_elevation_velocity,
        )


class IdealSensorTargetAtStarboardTestCase(absltest.TestCase):

    def setUp(self):
        # Configure the agent.
        agent_state = State()
        self.agent_velocity = np.array([0, 1, 0])
        (
            agent_state.velocity.x,
            agent_state.velocity.y,
            agent_state.velocity.z,
        ) = self.agent_velocity
        self.agent = ModelAgent(agent_state)
        self.sensor = IdealSensor(self.agent)

        # Configure the target.
        target_state = State()
        self.target_position = np.array([5, 0, 0])

        (
            target_state.position.x,
            target_state.position.y,
            target_state.position.z,
        ) = self.target_position
        self.target_velocity = np.array([2, 3, -1])
        (
            target_state.velocity.x,
            target_state.velocity.y,
            target_state.velocity.z,
        ) = self.target_velocity
        self.target = ModelAgent(target_state)

    def test_sense_position_range(self):
        expected_range = np.linalg.norm(self.target_position)
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.range,
            expected_range,
        )

    def test_sense_position_azimuth(self):
        expected_azimuth = np.arctan(self.target_position[0] /
                                     self.target_position[1])
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.azimuth,
            expected_azimuth,
        )

    def test_sense_position_elevation(self):
        expected_elevation = np.arctan(
            self.target_position[2] /
            np.sqrt(self.target_position[0]**2 + self.target_position[1]**2))
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.elevation,
            expected_elevation,
        )

    def test_sense_velocity_range(self):
        expected_range_rate = 2
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.range,
            expected_range_rate,
        )

    def test_sense_velocity_azimuth(self):
        expected_azimuth_velocity = -2 / 5
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.azimuth,
            expected_azimuth_velocity,
        )

    def test_sense_velocity_elevation(self):
        expected_elevation_velocity = -1 / 5
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.elevation,
            expected_elevation_velocity,
        )


class IdealSensorTargetAboveOnYawTestCase(absltest.TestCase):

    def setUp(self):
        # Configure the agent.
        agent_state = State()
        self.agent_velocity = np.array([0, 1, 0])
        (
            agent_state.velocity.x,
            agent_state.velocity.y,
            agent_state.velocity.z,
        ) = self.agent_velocity
        self.agent = ModelAgent(agent_state)
        self.sensor = IdealSensor(self.agent)

        # Configure the target.
        target_state = State()
        self.target_position = np.array([0, 0, 5])

        (
            target_state.position.x,
            target_state.position.y,
            target_state.position.z,
        ) = self.target_position
        self.target_velocity = np.array([0, 2, 0])
        (
            target_state.velocity.x,
            target_state.velocity.y,
            target_state.velocity.z,
        ) = self.target_velocity
        self.target = ModelAgent(target_state)

    def test_sense_position_range(self):
        expected_range = np.linalg.norm(self.target_position)
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.range,
            expected_range,
        )

    def test_sense_position_azimuth(self):
        expected_azimuth = 0
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.azimuth,
            expected_azimuth,
        )

    def test_sense_position_elevation(self):
        expected_elevation = np.arctan(
            self.target_position[2] /
            np.sqrt(self.target_position[0]**2 + self.target_position[1]**2))
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.elevation,
            expected_elevation,
        )

    def test_sense_velocity_range(self):
        expected_range_rate = 0
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.range,
            expected_range_rate,
        )

    def test_sense_velocity_azimuth(self):
        expected_azimuth_velocity = 0
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.azimuth,
            expected_azimuth_velocity,
        )

    def test_sense_velocity_elevation(self):
        expected_elevation_velocity = -1 / 5
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.elevation,
            expected_elevation_velocity,
        )


class IdealSensorTargetAtAzimuth45Elevation45TestCase(absltest.TestCase):

    def setUp(self):
        # Configure the agent.
        agent_state = State()
        self.agent_velocity = np.array([0, 1, 0])
        (
            agent_state.velocity.x,
            agent_state.velocity.y,
            agent_state.velocity.z,
        ) = self.agent_velocity
        self.agent = ModelAgent(agent_state)
        self.sensor = IdealSensor(self.agent)

        # Configure the target.
        target_state = State()
        self.target_position = np.array([1, 1, np.sqrt(2)])

        (
            target_state.position.x,
            target_state.position.y,
            target_state.position.z,
        ) = self.target_position
        self.target_velocity = (np.array([-1, 1, 0]) +
                                np.array([1, 1, -np.sqrt(2)]) +
                                self.agent_velocity)
        (
            target_state.velocity.x,
            target_state.velocity.y,
            target_state.velocity.z,
        ) = self.target_velocity
        self.target = ModelAgent(target_state)

    def test_sense_position_range(self):
        expected_range = np.linalg.norm(self.target_position)
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.range,
            expected_range,
        )

    def test_sense_position_azimuth(self):
        expected_azimuth = np.arctan(self.target_position[0] /
                                     self.target_position[1])
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.azimuth,
            expected_azimuth,
        )

    def test_sense_position_elevation(self):
        expected_elevation = np.arctan(
            self.target_position[2] /
            np.sqrt(self.target_position[0]**2 + self.target_position[1]**2))
        sensor_output = self.sensor.sense_position(self.target)
        self.assertAlmostEqual(
            sensor_output.position.elevation,
            expected_elevation,
        )

    def test_sense_velocity_range(self):
        expected_range_rate = 0
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.range,
            expected_range_rate,
        )

    def test_sense_velocity_azimuth(self):
        expected_azimuth_velocity = -np.sqrt(2) / 2
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.azimuth,
            expected_azimuth_velocity,
        )

    def test_sense_velocity_elevation(self):
        expected_elevation_velocity = -2 / 2
        sensor_output = self.sensor.sense_velocity(self.target)
        self.assertAlmostEqual(
            sensor_output.velocity.elevation,
            expected_elevation_velocity,
        )


if __name__ == "__main__":
    absltest.main()

File: simulation/swarm/sensors/ideal_sensor_test.cc
#include "simulation/swarm/sensors/ideal_sensor.h"

#include <gtest/gtest.h>

#include <Eigen/Dense>

#include "simulation/swarm/model_agent.h"
#include "simulation/swarm/proto/agent.pb.h"
#include "simulation/swarm/proto/state.pb.h"

namespace swarm::sensor {
namespace {

// Maximum error tolerance.
constexpr double kMaxErrorTolerance = 1e-6;

class IdealSensorTest : public testing::Test {
 protected:
  // Agent state.
  const Eigen::Vector3d kAgentPosition{0, 0, 0};
  const Eigen::Vector3d kAgentVelocity{0, 4, 0};

  // Target state.
  const Eigen::Vector3d kTargetPosition{0, 4, 0};
  const Eigen::Vector3d kTargetVelocity{2, 2, -1};

  IdealSensorTest() : sensor_(IdealSensor(agent_)) {
    State agent_state;
    agent_state.mutable_position()->set_x(kAgentPosition(0));
    agent_state.mutable_position()->set_y(kAgentPosition(1));
    agent_state.mutable_position()->set_z(kAgentPosition(2));
    agent_state.mutable_velocity()->set_x(kAgentVelocity(0));
    agent_state.mutable_velocity()->set_y(kAgentVelocity(1));
    agent_state.mutable_velocity()->set_z(kAgentVelocity(2));
    agent_ = agent::ModelAgent(std::move(agent_state));

    State target_state;
    target_state.mutable_position()->set_x(kTargetPosition(0));
    target_state.mutable_position()->set_y(kTargetPosition(1));
    target_state.mutable_position()->set_z(kTargetPosition(2));
    target_state.mutable_velocity()->set_x(kTargetVelocity(0));
    target_state.mutable_velocity()->set_y(kTargetVelocity(1));
    target_state.mutable_velocity()->set_z(kTargetVelocity(2));
    target_ = agent::ModelAgent(std::move(target_state));
  }

  // Ideal sensor.
  IdealSensor sensor_;

  // Agent.
  agent::ModelAgent agent_;

  // Target.
  agent::ModelAgent target_;
};

TEST_F(IdealSensorTest, SensePositionRange) {
  const auto expected_range = kTargetPosition.norm();
  const auto sensor_output = sensor_.SensePosition(target_);
  EXPECT_NEAR(sensor_output.position().range(), expected_range,
              kMaxErrorTolerance);
}

TEST_F(IdealSensorTest, SensePositionAzimuth) {
  const auto expected_azimuth =
      std::atan(kTargetPosition(0) / kTargetPosition(1));
  const auto sensor_output = sensor_.SensePosition(target_);
  EXPECT_NEAR(sensor_output.position().azimuth(), expected_azimuth,
              kMaxErrorTolerance);
}

TEST_F(IdealSensorTest, SensePositionElevation) {
  const auto expected_elevation = std::atan(
      kTargetPosition(2) / std::sqrt(std::pow(kTargetPosition(0), 2) +
                                     std::pow(kTargetPosition(1), 2)));
  const auto sensor_output = sensor_.SensePosition(target_);
  EXPECT_NEAR(sensor_output.position().elevation(), expected_elevation,
              kMaxErrorTolerance);
}

TEST_F(IdealSensorTest, SenseVelocityRange) {
  const auto expected_range_rate = -2.0;
  const auto sensor_output = sensor_.SenseVelocity(target_);
  EXPECT_NEAR(sensor_output.velocity().range(), expected_range_rate,
              kMaxErrorTolerance);
}

TEST_F(IdealSensorTest, SenseVelocityAzimuth) {
  const auto expected_azimuth_velocity = 2.0 / 4;
  const auto sensor_output = sensor_.SenseVelocity(target_);
  EXPECT_NEAR(sensor_output.velocity().azimuth(), expected_azimuth_velocity,
              kMaxErrorTolerance);
}

TEST_F(IdealSensorTest, SenseVelocityElevation) {
  const auto expected_elevation_velocity = -1.0 / 4;
  const auto sensor_output = sensor_.SenseVelocity(target_);
  EXPECT_NEAR(sensor_output.velocity().elevation(), expected_elevation_velocity,
              kMaxErrorTolerance);
}

}  // namespace
}  // namespace swarm::sensor
